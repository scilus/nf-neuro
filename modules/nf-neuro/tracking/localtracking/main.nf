process TRACKING_LOCALTRACKING {
    tag "$meta.id"
    label 'process_high_memory'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(wm), path(fodf), path(fa)

    output:
    tuple val(meta), path("*__local_tracking.trk"), emit: trk
    tuple val(meta), path("*__local_tracking_config.json"), emit: config
    tuple val(meta), path("*__local_seeding_mask.nii.gz"), emit: seedmask
    tuple val(meta), path("*__local_tracking_mask.nii.gz"), emit: trackmask
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def local_fa_tracking_mask_threshold = task.ext.local_fa_tracking_mask_threshold ? task.ext.local_fa_tracking_mask_threshold : ""
    def local_fa_seeding_mask_threshold = task.ext.local_fa_seeding_mask_threshold ? task.ext.local_fa_seeding_mask_threshold : ""
    def local_tracking_mask = task.ext.local_tracking_mask_type ? "${task.ext.local_tracking_mask_type}" : ""
    def local_seeding_mask = task.ext.local_seeding_mask_type ? "${task.ext.local_seeding_mask_type}" : ""

    def local_step = task.ext.local_step ? "--step " + task.ext.local_step : ""
    def local_random_seed = task.ext.local_random_seed ? "--seed " + task.ext.local_random_seed : ""
    def local_seeding = task.ext.local_seeding ? "--" + task.ext.local_seeding : ""
    def local_nbr_seeds = task.ext.local_nbr_seeds ? "" + task.ext.local_nbr_seeds : ""
    def local_min_len = task.ext.local_min_len ? "--min_length " + task.ext.local_min_len : ""
    def local_max_len = task.ext.local_max_len ? "--max_length " + task.ext.local_max_len : ""
    def local_theta = task.ext.local_theta ? "--theta "  + task.ext.local_theta : ""
    def local_sfthres = task.ext.local_sfthres ? "--sfthres "  + task.ext.local_sfthres : ""
    def local_algo = task.ext.local_algo ? "--algo " + task.ext.local_algo: ""
    def compress = task.ext.local_compress_streamlines ? "--compress " + task.ext.local_compress_value : ""
    def basis = task.ext.basis ? "--sh_basis " + task.ext.basis : ""

    def gpu_batch_size = task.ext.gpu_batch_size ? "--batch_size " + task.ext.gpu_batch_size : ""
    def enable_gpu = task.ext.enable_gpu ? "--use_gpu $gpu_batch_size" : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    if [ "${local_tracking_mask}" == "wm" ]; then
        scil_volume_math.py convert $wm ${prefix}__local_tracking_mask.nii.gz \
            --data_type uint8 -f

    elif [ "${local_tracking_mask}" == "fa" ]; then
        scil_volume_math.py lower_threshold $fa \
            $local_fa_tracking_mask_threshold \
            ${prefix}__local_tracking_mask.nii.gz \
            --data_type uint8 -f
    fi

    if [ "${local_seeding_mask}" == "wm" ]; then
        scil_volume_math.py convert $wm ${prefix}__local_seeding_mask.nii.gz \
            --data_type uint8 -f

    elif [ "${local_seeding_mask}" == "fa" ]; then
        scil_volume_math.py lower_threshold $fa \
            $local_fa_seeding_mask_threshold \
            ${prefix}__local_seeding_mask.nii.gz \
            --data_type uint8 -f
    fi

    scil_tracking_local.py $fodf ${prefix}__local_seeding_mask.nii.gz \
            ${prefix}__local_tracking_mask.nii.gz tmp.trk $enable_gpu\
            $local_algo $local_seeding $local_nbr_seeds\
            $local_random_seed $local_step $local_theta\
            $local_sfthres $local_min_len\
            $local_max_len $compress $basis -f

    scil_tractogram_remove_invalid.py tmp.trk\
            ${prefix}__local_tracking.trk\
            --remove_single_point -f

    cat <<-TRACKING_INFO > ${prefix}__local_tracking_config.json
    {"algorithm": "${task.ext.local_algo}",
    "fa_tracking_threshold": $task.ext.local_fa_tracking_mask_threshold,
    "fa_seeding_threshlod": $task.ext.local_fa_seeding_mask_threshold,
    "seeding_type": "${task.ext.local_seeding}",
    "tracking_mask": "${task.ext.local_tracking_mask_type}",
    "nb_seed": $task.ext.local_nbr_seeds,
    "seeding_mask": "${task.ext.local_seeding_mask_type}",
    "random_seed": $task.ext.local_random_seed,
    "is_compress": "${task.ext.local_compress_streamlines}",
    "compress_value": $task.ext.local_compress_value,
    "step": $task.ext.local_step,
    "theta": $task.ext.local_theta,
    "sfthres": $task.ext.local_sfthres,
    "min_len": $task.ext.local_min_len,
    "max_len": $task.ext.local_max_len,
    "sh_basis": "${task.ext.basis}"}
    TRACKING_INFO

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_tracking_local.py -h
    scil_tractogram_remove_invalid.py -h
    scil_volume_math.py -h

    touch ${prefix}__local_tracking.trk
    touch ${prefix}__local_tracking_config.json
    touch ${prefix}__local_seeding_mask.nii.gz
    touch ${prefix}__local_tracking_mask.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
