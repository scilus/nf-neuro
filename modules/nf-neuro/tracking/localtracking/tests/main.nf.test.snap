{
    "tracking - localtracking - wm": {
        "content": [
            "test__local_tracking.trk",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_tracking_config.json:md5,5eac37e6262a85d8d1b33c92e1f35796"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_seeding_mask.nii.gz:md5,6180aeda681e4443e56287ef1e1aee83"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_tracking_mask.nii.gz:md5,6180aeda681e4443e56287ef1e1aee83"
                ]
            ],
            [
                "versions.yml:md5,0088e7190e395324acbb8fce5c474750"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-24T19:38:16.835402429"
    },
    "tracking - localtracking - fa": {
        "content": [
            "test__local_tracking.trk",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_tracking_config.json:md5,a082f8d2f2b92cb3f2912a796942a786"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_seeding_mask.nii.gz:md5,0484a2e44ff153caa79819a8646cc33c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__local_tracking_mask.nii.gz:md5,0484a2e44ff153caa79819a8646cc33c"
                ]
            ],
            [
                "versions.yml:md5,0088e7190e395324acbb8fce5c474750"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-24T19:37:54.047881093"
    }
}